pkg/internal/cyberark/identity/identity.go

```diff
@@ -0,0 +1,441 @@
+	subdomain string
+
+	tokenCache      map[string]token
+	tokenCacheMutex sync.Mutex
+}
+
+// token is a wrapper type for holding auth tokens we want to cache.
+type token string
+
+// New returns an initialized CyberArk Identity client using a default service discovery client.
+// NB: This function performs service discovery when called, in order to ensure that all Identity
+// clients are created with a valid Identity API URL. This function blocks on the network call to
+// the discovery service.
+func New(ctx context.Context, subdomain string) (*Client, error) {
+	return NewWithDiscoveryClient(ctx, servicediscovery.New(), subdomain)
+}
+
+// NewWithDiscoveryClient returns an initialized CyberArk Identity client using the given service discovery client.
+// NB: This function performs service discovery when called, in order to ensure that all Identity
+// clients are created with a valid Identity API URL. This function blocks on the network call to
+// the discovery service.
+func NewWithDiscoveryClient(ctx context.Context, discoveryClient *servicediscovery.Client, subdomain string) (*Client, error) {
+	if discoveryClient == nil {
+		return nil, fmt.Errorf("must provide a non-nil discovery client to the Identity Client")
+	}
+
+	endpoint, err := discoveryClient.DiscoverIdentityAPIURL(ctx, subdomain)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		client: &http.Client{
+			Timeout: 10 * time.Second,
+		},
+
+		endpoint:  endpoint,
+		subdomain: subdomain,
+
+		tokenCache:      make(map[string]token),
+		tokenCacheMutex: sync.Mutex{},
+	}, nil
+}
+
+// LoginUsernamePassword performs a blocking call to fetch an auth token from CyberArk Identity using the given username and password.
+// The password is zeroed after use.
+// Tokens are cached internally and are not directly accessible to code; use Client.AuthenticatedHTTPClient to add credentials
+// to an *http.Client.
+func (c *Client) LoginUsernamePassword(ctx context.Context, username string, password []byte) error {
+	defer func() {
+		for i := range password {
+			password[i] = 0x00
+		}
+	}()
+
+	c.tokenCacheMutex.Lock()
+	defer c.tokenCacheMutex.Unlock()
+
+	advanceRequestBody, err := c.doStartAuthentication(ctx, username)
+	if err != nil {
+		return err
+	}
+
+	// We can't skip AdvanceAuthentication so we need to add the password to the body
+	// and send it off to complete the login process
+	advanceRequestBody.Answer = string(password)
+
+	err = c.doAdvanceAuthentication(ctx, username, &advanceRequestBody)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// doStartAuthentication performs the initial request to start the login process using a username and password.
+// It returns a partially initialized advanceAuthenticationRequestBody ready to send to the server to complete
+// the login. To avoid copying passwords, this function doesn't set the password in the Answer field, and
+// callers must set that field before making the request to AdvanceAuthentication.
```

Comment:
I don't understand the "avoid copying passwords" comment. This function doesn't have or need the password so why would it set it in the Answer field?
I would find it more useful to have a package level `doc.go` file or README or CONTRIBUTING doc, explaining how future maintainers should write code to minimise secrets getting copied and lingering in memory.

Reply:
I'll try to improve this, thanks!

Reply:
>  I would find it more useful to have a package level doc.go file or README or CONTRIBUTING doc, explaining how future maintainers should write code to minimise secrets getting copied and lingering in memory.

I'm gonna save this idea as something which would be useful for us later, but I'm not going to write this now. I've tried to improve the comments a bit for now!